ДЗ 2, Применение коллекций

anv0l opened the pull request:
Цель: попрактиковать различные аспекты коллекций.
познакомится с реализациями Map.
надо сдалать todo в классах из пакета homework.
Все тесты должны проходить.
Предполалается использование встроенного в jdk функционала, поэтому реализация методов должны быть буквально из нескольких строк.

anv0l committed:
ДЗ 2, Применение коллекций

Цель: попрактиковать различные аспекты коллекций.
познакомится с реализациями Map.
надо сдалать todo в классах из пакета homework.
Все тесты должны проходить.
Предполалается использование встроенного в jdk функционала, поэтому реализация методов должны быть буквально из нескольких строк.

petrelevich commented on hw02-generics/src/main/java/homework/Customer.java:
```diff
@@ -0,0 +1,75 @@
+    private long scores;
+
+    // Список уже добавленных клиентов
+    private static Map<Customer, Customer> customers = new HashMap<>();
```
Очень странное решение.
Почему  Customer содержит других Customer-ов?
не надо подгонять код под тесты.

anv0l replied:
Заменил HashMap на ArrayList. Не понимаю, как по-другому можно проверить при добавлении в коллекцию _нового_ объекта с ключом старого, что они одинаковые.

petrelevich replied:
Коллекция клиентов тут совсем не нужна.
Под какой тест сделано это решение?

anv0l replied:
Тест "Модификация коллекции".

petrelevich replied:
Получается, что как-то надо защитить содержимое коллекции от неожиданного изменения.
Например, можно возвращать копии элементов, чтобы оригинал остался прежним.

anv0l replied:
Про копии думал, но не понимал, как реализовать. Сделал с помощью new AbstractMap.SimpleEntry, тест проходит без ошибок, убрал из класса Customer лишнюю коллекцию.

petrelevich commented on hw02-generics/src/main/java/homework/Customer.java:
```diff
@@ -0,0 +1,75 @@
+        Customer customer = (Customer) o;
+
+        if (id != customer.id) return false;
+        if (scores != customer.scores) return false;
```
equals и HashCode должны по одним полям работать.

anv0l replied:
Изменил вычисление hashCode.

petrelevich commented on hw02-generics/src/main/java/homework/CustomerReverseOrder.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.*;
+
+public class CustomerReverseOrder {
+    private LinkedList<Customer> customers;
```
`Deque<Integer> stack = new ArrayDeque<Integer>();`
тут выглядит немного красивее.

anv0l replied:
Заменил на Deque.

petrelevich commented on hw02-generics/src/main/java/homework/CustomerService.java:
```diff
@@ -0,0 +1,60 @@
+    private Map<Customer, String> customers;
+
+    public CustomerService() {
+        customers = new HashMap<Customer, String>();
```
обратите внимание на подсказку про имплементацию Map.
Есть намного более простое решение, без Iterator.
Буквально в две строчки.

anv0l replied:
Переделал на TreeMap, стало действительно короче.

anv0l committed:
ДЗ 2, Применение коллекций

Цель: попрактиковать различные аспекты коллекций.
познакомится с реализациями Map.
надо сдалать todo в классах из пакета homework.
Все тесты должны проходить.
Предполалается использование встроенного в jdk функционала, поэтому реализация методов должны быть буквально из нескольких строк.

anv0l committed:
ДЗ 2, Применение коллекций

Цель: попрактиковать различные аспекты коллекций.
познакомится с реализациями Map.
надо сдалать todo в классах из пакета homework.
Все тесты должны проходить.
Предполалается использование встроенного в jdk функционала, поэтому реализация методов должны быть буквально из нескольких строк.

anv0l commented:
Также обновил тест.

anv0l committed:
ДЗ 2, Применение коллекций

Цель: попрактиковать различные аспекты коллекций.
познакомится с реализациями Map.
надо сдалать todo в классах из пакета homework.
Все тесты должны проходить.
Предполалается использование встроенного в jdk функционала, поэтому реализация методов должны быть буквально из нескольких строк.

petrelevich commented on hw02-generics/src/main/java/homework/CustomerReverseOrder.java:
```diff
@@ -0,0 +1,24 @@
+import java.util.*;
+
+public class CustomerReverseOrder {
+    private Deque<Customer> customers;
```
стоит добавить final

petrelevich commented on hw02-generics/src/main/java/homework/CustomerService.java:
```diff
@@ -0,0 +1,45 @@
+    private TreeMap<Customer, String> customers;
+
+    public CustomerService() {
+        customers = new TreeMap<>(new Comparator<>() {
```
можно упростить customers = new TreeMap<>(Comparator.comparing(Customer::getScores));

petrelevich commented on hw02-generics/src/main/java/homework/CustomerService.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.*;
+
+public class CustomerService {
+    private TreeMap<Customer, String> customers;
```
лучше:
`private final NavigableMap<Customer, String> customers;`

petrelevich commented on hw02-generics/src/main/java/homework/CustomerService.java:
```diff
@@ -0,0 +1,45 @@
+
+    public Map.Entry<Customer, String> getNext(Customer customer) {
+        for (Map.Entry<Customer, String> entryCustomer : this.customers.entrySet()) {
+            if (entryCustomer.getKey().getScores() > customer.getScores() ) {
```
можно проще так:
```
    public Map.Entry<Customer, String> getNext(Customer customer) {
        Map.Entry<Customer, String> higherEntry = customers.higherEntry(customer);
        return higherEntry == null ? null : new CustomerEntry(higherEntry.getKey(), higherEntry.getValue());
    }
```

CustomerEntry - это 
` private static class CustomerEntry implements Map.Entry<Customer,String>`

anv0l committed:
ДЗ 2, Применение коллекций

Цель: попрактиковать различные аспекты коллекций.
познакомится с реализациями Map.
надо сдалать todo в классах из пакета homework.
Все тесты должны проходить.
Предполалается использование встроенного в jdk функционала, поэтому реализация методов должны быть буквально из нескольких строк.
